pkg/dockerregistry/server/signaturedispatcher.go

```diff
@@ -0,0 +1,162 @@
+package server
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	kapierrors "k8s.io/kubernetes/pkg/api/errors"
+
+	ctxu "github.com/docker/distribution/context"
+
+	"github.com/docker/distribution/context"
+	"github.com/docker/distribution/registry/api/errcode"
+	"github.com/docker/distribution/registry/api/v2"
+	"github.com/docker/distribution/registry/handlers"
+
+	imageapi "github.com/openshift/origin/pkg/image/api"
+
+	gorillahandlers "github.com/gorilla/handlers"
+)
+
+type signature struct {
+	Name    string `json:"name"`
+	Type    string `json:"type"`
+	Content []byte `json:"content"`
+}
+
+type signatureList struct {
+	Signatures []signature `json:"signatures"`
+}
+
+const errGroup = "registry.api.v2"
+
+var (
+	ErrorCodeSignatureInvalid = errcode.Register(errGroup, errcode.ErrorDescriptor{
+		Value:          "SIGNATURE_INVALID",
+		Message:        "invalid image signature",
+		HTTPStatusCode: http.StatusBadRequest,
+	})
+
+	ErrorCodeSignatureAlreadyExists = errcode.Register(errGroup, errcode.ErrorDescriptor{
+		Value:          "SIGNATURE_EXISTS",
+		Message:        "image signature already exists",
+		HTTPStatusCode: http.StatusBadRequest,
+	})
+)
+
+// SignatureDispatcher dispatch the signatures endpoint.
+func SignatureDispatcher(ctx *handlers.Context, r *http.Request) http.Handler {
+	signatureHandler := &signatureHandler{
+		Context: ctx,
+	}
+	signatureHandler.Reference, _ = imageapi.ParseDockerImageReference(ctxu.GetStringValue(ctx, "vars.name") + "@" + ctxu.GetStringValue(ctx, "vars.digest"))
+	return gorillahandlers.MethodHandler{
+		"GET": http.HandlerFunc(signatureHandler.Get),
+		"PUT": http.HandlerFunc(signatureHandler.Put),
+	}
+}
+
+type signatureHandler struct {
+	Context   *handlers.Context
```

Comment:
@mfojtik The rest of our code usually call this variable as ctx. @miminar mentioned it.